### PR TITLE
Put Pin::as_deref_mut in impl Pin<Ptr> / rearrange Pin methods

### PR DESCRIPTION
**Pull request by coolreader18**
Tracking issue: #86918

Based on the suggestion in https://github.com/rust-lang/rust/issues/86918#issuecomment-2189367582

> Some advantages:
> 
>  * Synergy with the existing `as_ref` and `as_mut` signatures (stable since Rust 1.33)
> 
>  * Lifetime elision reduces noise in the signature
> 
>  * Turbofish less verbose: `Pin::<&mut T>::as_deref_mut` vs `Pin::<&mut Pin<&mut T>>::as_deref_mut`

The comment seemed to imply that `Pin::as_ref` and `Pin::as_mut` already share an impl block, which they don't. So, I rearranged it so that they do, and we can see which looks better in the docs.

<details><summary><b>Docs screenshots</b></summary>

Current nightly:
image

`Pin::as_deref_mut` moved into the same block as `as_mut`:
image

`Pin::as_ref`, `as_mut`, and `as_deref_mut` all in the same block:
image

</details>

I think I like the last one the most; obviously I'm biased since I'm the one who rearranged it, but it doesn't make sense to me to have `as_ref` methods split up by an `into_inner` method.

r? dtolnay

### CODE REVIEW

File: library/core/src/pin.rs [hunk 1]
```diff
@@ -1292,7 +1292,7 @@ impl<Ptr: Deref> Pin<Ptr> {
     ///     // data that we pinned above, which we could use to move it as we have
     ///     // seen in the previous example. We have violated the pinning API contract.
     ///  }
-    ///  ```
+    /// ```
```
Comment (dtolnay) [inline library/core/src/pin.rs:1295]: If fixing this, let's also fix the previous line which is equally misaligned.

**Review by dtolnay — COMMENTED** [review]
Thank you!

**Comment by coolreader18** [comment]
I guess the other option is to put `into_inner_unchecked` at the end of the block - it's certainly less used than the other methods would be, but then that messes up the synchrony with the `new()` & `into_inner()` methods being next to each other for `Unpin` types. Though that is how it is currently anyway. I'm gonna try that.

**Comment by coolreader18** [comment]
`Pin::new_unchecked` followed directly by the `as_ref` methods:
image

**Review by dtolnay — APPROVED** [review]

**Comment by ogoffart** [comment]
Btw, the changes of bounds in Pin::set from this PR most likely causes regression described in https://github.com/rust-lang/rust/issues/129601

**Comment by coolreader18** [comment]
Ah, shoot, that makes sense.

**Pull request merged**